authentication/grpc.go

```diff
@@ -0,0 +1,91 @@
+package authentication
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/go-kit/log"
+	grpc_middleware "github.com/grpc-ecosystem/go-grpc-middleware/v2"
+	grpc_middleware_auth "github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/auth"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+// GRPCMiddlewareFunc is a function type able to return authentication middleware for
+// a given tenant. If no middleware is found, the second return value should be false.
+type GRPCMiddlewareFunc func(tenant string) (grpc.StreamServerInterceptor, bool)
+
+// WithGRPCTenantHeader returns a new StreamServerInterceptor that adds the tenant and tenantid
+// to the stream Context.
+func WithGRPCTenantHeader(header string, tenantIDs map[string]string, logger log.Logger) grpc.StreamServerInterceptor {
+	header = strings.ToLower(header)
+
+	return func(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		ctx := ss.Context()
+		md, ok := metadata.FromIncomingContext(ctx)
+		if !ok {
+			return status.Errorf(codes.Internal, "metadata not found")
+		}
+
+		headerTenants := md[header]
+		if len(headerTenants) == 0 {
+			return status.Errorf(codes.InvalidArgument, fmt.Sprintf("header %q not set", header))
+		}
+		if len(headerTenants) > 1 {
+			return status.Errorf(codes.InvalidArgument, fmt.Sprintf("header %q requested multiple tenants", header))
+		}
+
+		tenant := headerTenants[0]
+		ctx = context.WithValue(ctx, tenantKey, tenant)
+
+		id, ok := tenantIDs[tenant]
+		if !ok {
+			// Rather than letting the caller know their tenant was invalid, just record that it was invalid
+			id = "__no_tenant_id"
+		}
+		ctx = context.WithValue(ctx, tenantIDKey, id)
+
+		wrapped := grpc_middleware.WrapServerStream(ss)
+		wrapped.WrappedContext = ctx
+
+		return handler(srv, wrapped)
+	}
+}
+
+func WithGRPCAccessToken() grpc.StreamServerInterceptor {
+	return grpc_middleware_auth.StreamServerInterceptor(func(ctx context.Context) (context.Context, error) {
+		md, ok := metadata.FromIncomingContext(ctx)
+		if !ok {
+			return nil, status.Error(codes.Internal, "metadata error")
```

Comment:
Nit: perhaps we could align the error message with the other interceptor (if we switch to 'not found' code there, it should also apply to this)